golang/cosmos/x/swingset/types/msgs.go

```diff
@@ -330,13 +330,82 @@ func (msg MsgInstallBundle) ValidateBasic() error {
 	if msg.Submitter.Empty() {
 		return sdkerrors.Wrap(sdkerrors.ErrInvalidAddress, "Submitter address cannot be empty")
 	}
-	if len(msg.Bundle) == 0 {
+	if len(msg.Bundle) == 0 && len(msg.CompressedBundle) == 0 {
 		return sdkerrors.Wrap(sdkerrors.ErrUnknownRequest, "Bundle cannot be empty")
 	}
+	if len(msg.Bundle) != 0 && len(msg.CompressedBundle) != 0 {
+		return sdkerrors.Wrap(sdkerrors.ErrUnknownRequest, "Cannot submit both compressed and uncompressed bundles")
+	}
+	if len(msg.Bundle) > 0 && msg.UncompressedSize != 0 {
+		return sdkerrors.Wrap(sdkerrors.ErrUnknownRequest, "Uncompressed size not be set when not compressing")
+	}
+	if len(msg.CompressedBundle) > 0 && !(msg.UncompressedSize > 0) {
+		return sdkerrors.Wrap(sdkerrors.ErrUnknownRequest, "Uncompressed size must be positive")
+	}
+	// We don't check the accuracy of the uncompressed size here, since it could comsume significant CPU.
 	return nil
 }
 
 // GetSigners defines whose signature is required
 func (msg MsgInstallBundle) GetSigners() []sdk.AccAddress {
 	return []sdk.AccAddress{msg.Submitter}
 }
+
+// ExpectedUncompressedSize returns the expected uncompressed size of the bundle.
+func (msg MsgInstallBundle) ExpectedUncompressedSize() uint64 {
+	if msg.UncompressedSize > 0 {
+		return uint64(msg.UncompressedSize)
+	}
+	return uint64(len(msg.Bundle))
+}
+
+// Compress gzip-compresses a validated bundle.
+// Do nothing if it looks like it's already compressed.
+func (msg *MsgInstallBundle) Compress() error {
+	if len(msg.Bundle) == 0 {
+		return nil
+	}
+	msg.UncompressedSize = int64(len(msg.Bundle))
+	inBuf := strings.NewReader(msg.Bundle)
+	var outBuf bytes.Buffer
+	gzipWriter := gzip.NewWriter(&outBuf)
+	_, err := io.Copy(gzipWriter, inBuf)
+	if err != nil {
+		return err
+	}
+	gzipWriter.Close() // required to flush to underlying buffer
+	msg.CompressedBundle = outBuf.Bytes()
+	msg.Bundle = ""
+	return nil
+}
+
+// Uncompress gzip-uncompresses a validated bundle.
+// Do nothing if it looks like it's already uncompressed.
+// Returns an error (and ends uncompression early) if the uncompressed
+// size does not match the expected uncompressed size.
+func (msg *MsgInstallBundle) Uncompress() error {
+	if len(msg.Bundle) > 0 {
+		return nil
+	}
+	bytesReader := bytes.NewReader(msg.CompressedBundle)
+	ungzipReader, err := gzip.NewReader(bytesReader)
+	if err != nil {
+		return err
+	}
+	limitedReader := io.LimitedReader{R: ungzipReader, N: msg.UncompressedSize}
+	var buf bytes.Buffer
+	n, err := io.Copy(&buf, &limitedReader)
+	if err != nil {
+		return err
+	}
+	if n != msg.UncompressedSize {
+		return sdkerrors.Wrap(sdkerrors.ErrInvalidRequest, "Uncompressed size does not match expected value")
+	}
+	if limitedReader.N > 0 {
+		return sdkerrors.Wrap(sdkerrors.ErrInvalidRequest, "Uncompressed size larger thah expected value")
```

Comment:
```suggestion
		return sdkerrors.Wrap(sdkerrors.ErrInvalidRequest, "Uncompressed size larger than expected value")
```

Reply:
Done. Someone must have been coding this in the wee hours.